pkg/server/registry/apigroups/acorn/events/strategy.go

```diff
@@ -53,28 +67,107 @@ type query struct {
 	tail int64
 }
 
-func (q query) on(list *apiv1.EventList) (*apiv1.EventList, error) {
-	// Sort in descending order
-	items := list.Items
-	sort.Slice(items, func(i, j int) bool {
-		return items[i].Observed.After(items[j].Observed.Time)
+func (q query) filterList(list *apiv1.EventList) *apiv1.EventList {
+	list.Items = q.filter(list.Items...)
+	return list
+}
+
+func (q query) filterEvent(e watch.Event) *watch.Event {
+	switch e.Type {
+	case watch.Added, watch.Modified:
+	default:
+		// Return unmodified
+		return &e
+	}
+
+	// Attempt to filter
+	obj := e.Object.(*apiv1.Event)
+	filtered := q.filter(*obj)
+	if len(filtered) < 1 {
+		// Drop the event, it's been filtered out
+		return nil
+	}
+
+	e.Object = filtered[0].DeepCopy()
+
+	return &e
+}
+
+func (q query) filter(events ...apiv1.Event) []apiv1.Event {
+	// Sort in chronological order
+	sort.Slice(events, func(i, j int) bool {
+		return events[i].Observed.Before(events[j].Observed.Time)
 	})
 
-	tail := len(items)
+	tail := len(events)
 	if q.tail > 0 && q.tail < int64(tail) {
 		tail = int(q.tail)
 	}
 
-	list.Items = items[:tail]
+	events = events[len(events)-tail:]
 
 	if q.details {
-		return list, nil
+		return events
 	}
 
-	for i, event := range list.Items {
+	for i, event := range events {
 		event.Details = nil
-		list.Items[i] = event
+		events[i] = event
+	}
+
+	return events
+}
+
+// stripQuery extracts the query from the given options, returning the query
+// and new options sans the query.
+func stripQuery(opts storage.ListOptions) (q query, stripped storage.ListOptions, err error) {
+	stripped = opts
+
+	stripped.Predicate.Field, err = stripped.Predicate.Field.Transform(func(f, v string) (string, string, error) {
+		var err error
+		switch f {
+		case "details":
+			q.details, err = strconv.ParseBool(v)
+		default:
+			return f, v, nil
+		}
+
+		return "", "", err
+	})
+	if err != nil {
+		return
+	}
+
+	q.tail, stripped.Predicate.Limit = stripped.Predicate.Limit, 0
+
+	return
+}
+
+func forward(ctx context.Context, from <-chan watch.Event, to chan<- watch.Event, q query) error {
```

Comment:
Can we add header comments to this and the other functions?

Reply:
done